### PR TITLE
io/lecture: Add second part of lecture (IPC)

### PR DESCRIPTION
**Pull request by Alex-deVis**
Add second part of the I/O lecture: Interprocess communication interface.

### CODE REVIEW

File: content/chapters/io/lecture/slides/IPC.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+### Client and Server
+
+- A process that use the socket interface is either a client or a server
+  - The server waits for messages/connections on IP and port
+  - The client attempts to connect to a server using IP and port
+
+----
+
+### TCP and UDP
```
Comment (razvand) [inline content/chapters/io/lecture/slides/IPC.md:95]: I would use the terms `Stream` and `Datagram`, and map them to `SOCK_STREAM` and `SOCK_DGRAM`.

File: content/chapters/io/lecture/slides/IPC.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+----
+
+#### TCP - `recv()`
+
+- Use socket to receive bytes
+- Return number of received bytes
+  - **-1** - recv failed
+  - **0** - connection end (similar to EOF)
+  - **<num** - partial recv
```
Comment (razvand) [inline content/chapters/io/lecture/slides/IPC.md:189]: I would add one `Summary` slide with two columns: `Server` and `Client`. Animated, with the common sequence of actions happening:

1. server creates `socket()`.
1. server `bind()` socket
1. server `listen()` on socket
1. server `accept()` on socket (and blocks)
1. client creates `socket()`
1. (optional for `SOCK_STREAM`) client binds `socket()`
1. client `connect()` to server
1. server `accept()` call returns a new socket
1. server `recv()` request from client
1. client `send()` request to server
1. client `recv()` reply from server
1. server `send()` reply to client
1. client `close()` connection to server
1. server `close()` connection to client

File: content/chapters/io/lecture/slides/devices.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+---
+
+### Device Types
+
+| Char Devices          | Block Devices             |
+| :-------------------: | :-----------------------: |
+| read/write one byte   | read/write blocks of data |
+| slow                  | fast                      |
+| no search             | search                    |
```
Comment (razvand) [inline content/chapters/io/lecture/slides/devices.md:36]:
```suggestion
| no seek             | seek                    |
```

File: content/chapters/io/lecture/slides/devices.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+| no buffering          | buffering                 |
+
```
Comment (razvand) [inline content/chapters/io/lecture/slides/devices.md:38]: I would give some `dd` examples that show how information is "stored" for block devices and any time you access it you get the same data. Unlike a char device such as `/dev/urandom`.

**Review by razvand — CHANGES_REQUESTED** [review]
This is great work, @Alex-deVis , thanks for this.

I have some small comments. And adding demos with sockets: UNIX and network, stream and datagram. You can take inspiration from the [receive challenges in the I/O lab](https://github.com/open-education-hub/operating-systems/tree/master/content/chapters/io/lab/solution/receive-challenges).

File: content/chapters/io/lecture/slides/IPC.md (hunk 1 above)
Comment (Alex-deVis) [reply content/chapters/io/lecture/slides/IPC.md]: Agreed.

File: content/chapters/io/lecture/slides/devices.md (hunk 4 above)
Comment (Alex-deVis) [reply content/chapters/io/lecture/slides/devices.md:38]: I went with the first approach.

**Review by teodutu — APPROVED** [review]

File: content/chapters/io/lecture/slides/IPC.md [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+
+### `mkfifo()` demo
+
+- `demo/IPC/fifo.c`
+
+---
+
+### IPC - Socket Interface
+
+![Socket Interface](../media/socket-interface.svg)
```
Comment (razvand) [inline content/chapters/io/lecture/slides/IPC.md:97]:
```suggestion
![Socket Interface](./media/socket-interface.svg)
```

File: content/chapters/io/lecture/slides/IPC.md [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+- The endpoint in the interprocess communication
+- Uniquely identifies the process in the communication
+- Supports multiple transmission types
+  - e.g.: `stream`, `datagram`
+
+----
+
+### Socket Interface - `socket()`
+
+![socket()](../media/socket-interface-socket.svg)
```
Comment (razvand) [inline content/chapters/io/lecture/slides/IPC.md:112]:
```suggestion
![socket()](./media/socket-interface-socket.svg)
```

**Review by razvand — CHANGES_REQUESTED** [review]
Please create two commits. One for `IPC` and one for `devices`. Prefix each commit message with `io/lecture:` And add a detailed description of each commit (something more than just the first line).

Also, there are some imperfections at the socket summary slide. There seems to be a bit of image-up, a bit of image-down. If you could fix that, it would be great. But it's not critical, we can merge this and fix it later on.

**Comment by razvand** [comment]
@Alex-deVis , also, please only use Tab-indenting. Not sure why this is not captured by the linter. @teodutu , any idea? The linter should check whether indenting is only done using Tabs.

**Comment by razvand** [comment]
> @Alex-deVis , also, please only use Tab-indenting. Not sure why this is not captured by the linter. @teodutu , any idea? The linter should check whether indenting is only done using Tabs.

@Alex-deVis , I created [a new commit](https://github.com/open-education-hub/operating-systems/pull/43/commits/a0a683d2a41273c99f479d9a71f354146b33ab74) with the source code Tab-indented.

**Comment by razvand** [comment]
@Alex-deVis , the Tab-indentation is only for **code**, not for Markdown.

**Review by razvand — APPROVED** [review]
All good.

**Pull request merged**